core/services/ocr2/plugins/vault/plugin.go

```diff
@@ -2,22 +2,756 @@ package vault
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+	"github.com/smartcontractkit/libocr/quorumhelper"
+	"github.com/smartcontractkit/tdh2/go/tdh2/tdh2easy"
+	"google.golang.org/protobuf/proto"
 
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/actions/vault"
 	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/requests"
 )
 
-func NewReportingPluginFactory(store *requests.Store[*Request]) *ReportingPluginFactory {
+const (
+	defaultBatchSize = 20
+	defaultNamespace = "main"
+	keySeparator     = ":"
+)
+
+var (
+	isValidIDComponent = regexp.MustCompile(`^[a-zA-Z0-9_]+$`).MatchString
+)
+
+func NewReportingPluginFactory(lggr logger.Logger, store *requests.Store[*Request], publicKey *tdh2easy.PublicKey, privateKeyShare *tdh2easy.PrivateShare) *ReportingPluginFactory {
 	return &ReportingPluginFactory{
-		store: store,
+		lggr:            lggr,
+		store:           store,
+		batchSize:       defaultBatchSize, // TODO fetch from onchain config
+		publicKey:       publicKey,
+		privateKeyShare: privateKeyShare,
 	}
 }
 
 type ReportingPluginFactory struct {
-	store *requests.Store[*Request]
+	lggr            logger.Logger
+	store           *requests.Store[*Request]
+	batchSize       int
+	publicKey       *tdh2easy.PublicKey
+	privateKeyShare *tdh2easy.PrivateShare
+}
+
+func (r *ReportingPluginFactory) NewReportingPlugin(ctx context.Context, config ocr3types.ReportingPluginConfig, fetcher ocr3_1types.BlobBroadcastFetcher) (ocr3_1types.ReportingPlugin[[]byte], ocr3_1types.ReportingPluginInfo, error) {
+	return &ReportingPlugin{
+		lggr:            r.lggr.Named("ReportingPlugin"),
+		store:           r.store,
+		batchSize:       r.batchSize,
+		config:          config,
+		publicKey:       r.publicKey,
+		privateKeyShare: r.privateKeyShare,
+	}, ocr3_1types.ReportingPluginInfo{}, nil
+}
+
+type ReportingPlugin struct {
+	lggr                           logger.Logger
+	store                          *requests.Store[*Request]
+	batchSize                      int
+	config                         ocr3types.ReportingPluginConfig
+	publicKey                      *tdh2easy.PublicKey
+	privateKeyShare                *tdh2easy.PrivateShare
+	maxSecretsPerOwner             int
+	maxCiphertextLenBytes          int
+	maxIdentifierKeyLenBytes       int
+	maxIdentifierOwnerLenBytes     int
+	maxIdentifierNamespaceLenBytes int
+}
+
+func (r *ReportingPlugin) Query(ctx context.Context, seqNr uint64, keyValueReader ocr3_1types.KeyValueReader, blobBroadcastFetcher ocr3_1types.BlobBroadcastFetcher) (types.Query, error) {
+	return types.Query{}, nil
+}
+
+func (r *ReportingPlugin) Observation(ctx context.Context, seqNr uint64, aq types.AttributedQuery, keyValueReader ocr3_1types.KeyValueReader, blobBroadcastFetcher ocr3_1types.BlobBroadcastFetcher) (types.Observation, error) {
+	// Note: this could mean that we end up processing more than `batchSize` requests
```

Comment:
nit: In general, a better strategy is to pack requests up to max Observation size (in bytes) configured for libOCR. Maciej is adding this for our main plugin. For now it shouldn't matter too much.